cyral/resource_cyral_repository_user_account.go

```diff
@@ -0,0 +1,596 @@
+		}
+	} else if resource.AuthScheme.CyralStorage != nil {
+		authScheme = []interface{}{
+			map[string]interface{}{
+				"cyral_storage": []interface{}{
+					map[string]interface{}{
+						"password": resource.AuthScheme.CyralStorage.Password,
+					},
+				},
+			},
+		}
+	} else if resource.AuthScheme.EnvironmentVariable != nil {
+		authScheme = []interface{}{
+			map[string]interface{}{
+				"environment_variable": []interface{}{
+					map[string]interface{}{
+						"variable_name": resource.AuthScheme.EnvironmentVariable.VariableName,
+					},
+				},
+			},
+		}
+	} else if resource.AuthScheme.GCPSecretManager != nil {
+		authScheme = []interface{}{
+			map[string]interface{}{
+				"gcp_secrets_manager": []interface{}{
+					map[string]interface{}{
+						"secret_name": resource.AuthScheme.GCPSecretManager.SecretName,
+					},
+				},
+			},
+		}
+	} else if resource.AuthScheme.HashicorpVault != nil {
+		authScheme = []interface{}{
+			map[string]interface{}{
+				"hashicorp_vault": []interface{}{
+					map[string]interface{}{
+						"path":                    resource.AuthScheme.HashicorpVault.Path,
+						"is_dynamic_user_account": resource.AuthScheme.HashicorpVault.IsDynamicUserAccount,
+					},
+				},
+			},
+		}
+	} else if resource.AuthScheme.KubernetesSecret != nil {
+		authScheme = []interface{}{
+			map[string]interface{}{
+				"kubernetes_secret": []interface{}{
+					map[string]interface{}{
+						"secret_name": resource.AuthScheme.KubernetesSecret.SecretName,
+						"secret_key":  resource.AuthScheme.KubernetesSecret.SecretKey,
+					},
+				},
+			},
+		}
+	} else {
+		return fmt.Errorf("auth scheme is required, user account is corrupt: %v", resource)
+	}
+
+	if err := d.Set("auth_scheme", authScheme); err != nil {
+		return fmt.Errorf("error setting 'auth_scheme': %w", err)
+	}
+
+	return nil
+}
+
+// ReadFromSchema is used to translate a .tf file into whatever the
+// UserAccounts API expects.
+func (userAccount *UserAccountResource) ReadFromSchema(d *schema.ResourceData) error {
+	// Set basic values required fields (UserID is computed).
+	userAccount.Name = d.Get("name").(string)
+	userAccount.AuthDatabaseName = d.Get("auth_database_name").(string)
+
+	// Handle approval config (optional field).
+	approvalConfig := d.Get("approval_config").(*schema.Set).List()
+	if len(approvalConfig) > 0 {
+		m := approvalConfig[0].(map[string]interface{})
+		userAccount.Config = &UserAccountConfig{
+			Approval: &ApprovalConfig{},
+		}
+		autogrant, ok := m["automatic_grant"]
+		if ok {
+			userAccount.Config.Approval.AutomaticGrant =
+				autogrant.(bool)
+		}
+		maxAutoGrant, ok := m["max_auto_grant_duration"]
+		if ok {
+			userAccount.Config.Approval.MaxAutomaticGrantDuration =
+				maxAutoGrant.(string)
+		}
+	}
+
+	// Handle Auth Scheme (required field).
+	authSchemeSet := d.Get("auth_scheme").(*schema.Set).List()
+	if len(authSchemeSet) != 1 {
+		return fmt.Errorf(
+			"exactly one auth_scheme attribute is required",
+		)
+	}
```

Comment:
This is already validated by the Required: true in the `auth_scheme`.
```suggestion
```

Reply:
Yes, I added a bunch of error catching while developing to make sure I could figure out what is going on. Do you think I should remove these cases, since we know that there should be no error? Just to make it clear when reading the code what is expected.

Reply:
Personally, Im usually in favor of removing this type of validation if its already being done in the schema since I think defining the validation in the schema is more "terraform idiomatic". But this is just an opinion that I have, both options will work in the end so feel free to decide about that